mlir/lib/Dialect/PDL/IR/Builtins.cpp

```diff
@@ -0,0 +1,56 @@
+#include <mlir/Dialect/PDL/IR/Builtins.h>
+#include <mlir/IR/PatternMatch.h>
+
+using namespace mlir;
+
+namespace mlir::pdl {
+namespace builtin {
+mlir::Attribute createDictionaryAttr(mlir::PatternRewriter &rewriter) {
+  return rewriter.getDictionaryAttr({});
+}
+
+mlir::Attribute addEntryToDictionaryAttr(mlir::PatternRewriter &rewriter,
+                                         mlir::Attribute dictAttr,
+                                         mlir::Attribute attrName,
+                                         mlir::Attribute attrEntry) {
+  assert(isa<DictionaryAttr>(dictAttr));
+  auto attr = dictAttr.cast<DictionaryAttr>();
+  auto name = attrName.cast<StringAttr>();
+  std::vector<NamedAttribute> values = attr.getValue().vec();
+
+  // Remove entry if it exists in the dictionary.
+  llvm::erase_if(values, [&](NamedAttribute &namedAttr) {
+    return namedAttr.getName() == name.getValue();
+  });
+
+  values.push_back(rewriter.getNamedAttr(name, attrEntry));
+  return rewriter.getDictionaryAttr(values);
+}
+
+mlir::Attribute createArrayAttr(mlir::PatternRewriter &rewriter) {
+  return rewriter.getArrayAttr({});
+}
+
+mlir::Attribute addElemToArrayAttr(mlir::PatternRewriter &rewriter,
+                                   mlir::Attribute attr,
+                                   mlir::Attribute element) {
```

Comment:
I am wondering if it's a good idea to change these native signature to the generic kind (`LogicalResult (PatternRewriter &, PDLResultList &, ArrayRef<PDLValue>)`) so that you can register it them as constraint with results?
or you prefer to register them with a lambda?

Reply:
Your comment has merit, and I think we will/should eventually have built-in constraints.

I just moved those functions from our internal repository.
The downside of switching native rewrites to that signature is that they come less readable and need extra code to handle the PDLResultList and ArrayRef<PDLValue>.

So I suggest we come back to native constraints when we need them, what do you think?